homeassistant/components/mobile_app/webhook.py

```diff
@@ -0,0 +1,174 @@
+                    CONF_SECRET, CONF_USER_ID, DOMAIN,
+                    HTTP_X_CLOUD_HOOK_ID, HTTP_X_CLOUD_HOOK_URL,
+                    INTEGRATIONS_MAP,
+                    WEBHOOK_PAYLOAD_SCHEMA, WEBHOOK_SCHEMAS, WEBHOOK_TYPES,
+                    WEBHOOK_TYPE_CALL_SERVICE, WEBHOOK_TYPE_FIRE_EVENT,
+                    WEBHOOK_TYPE_RENDER_TEMPLATE, WEBHOOK_TYPE_UPDATE_LOCATION,
+                    WEBHOOK_TYPE_UPDATE_REGISTRATION)
+
+from .helpers import (device_context, _decrypt_payload, empty_okay_response,
+                      safe_device, savable_state)
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def register_device_webhook(hass: HomeAssistantType, store: Store,
+                            device: Dict) -> None:
+    """Register the webhook for a device."""
+    device_name = 'Mobile App: {}'.format(device[ATTR_DEVICE_NAME])
+    webhook_id = device[CONF_WEBHOOK_ID]
+    webhook_register(hass, DOMAIN, device_name, webhook_id,
+                     partial(handle_webhook, store))
+
+
+async def handle_webhook(store: Store, hass: HomeAssistantType,
+                         webhook_id: str, request: Request) -> Response:
+    """Handle webhook callback."""
+    if webhook_id in hass.data[DOMAIN][ATTR_DELETED_IDS]:
+        return Response(status=410)
+
+    device = hass.data[DOMAIN][ATTR_REGISTRATIONS][webhook_id]
+
+    try:
+        req_data = await request.json()
+    except ValueError:
+        _LOGGER.warning('Received invalid JSON from mobile_app')
+        return json_response([], status=HTTP_BAD_REQUEST)
+
+    try:
+        req_data = WEBHOOK_PAYLOAD_SCHEMA(req_data)
+    except vol.Invalid as ex:
+        err = vol.humanize.humanize_error(req_data, ex)
+        _LOGGER.error('Received invalid webhook payload: %s', err)
+        return empty_okay_response()
+
+    headers = {}
+
+    via_cloud = is_cloudhook_request(request)
+
+    logged_in = async_is_logged_in(hass)
+
+    if via_cloud is False and logged_in is True:
+        cloud_id = device.get(CONF_CLOUDHOOK_ID)
+        cloud_url = device.get(CONF_CLOUDHOOK_URL)
+        if cloud_url is None:
+            try:
+                hook = await async_create_cloudhook(hass, webhook_id)
+                cloud_id = hook[CONF_CLOUDHOOK_ID]
+                cloud_url = hook[CONF_CLOUDHOOK_URL]
+            except CloudNotAvailable:
+                _LOGGER.error("Error creating cloudhook during local call")
+
+        headers[HTTP_X_CLOUD_HOOK_ID] = cloud_id
+        headers[HTTP_X_CLOUD_HOOK_URL] = cloud_url
+
+    webhook_type = req_data[ATTR_WEBHOOK_TYPE]
+
+    webhook_payload = req_data.get(ATTR_WEBHOOK_DATA, {})
+
+    if req_data[ATTR_WEBHOOK_ENCRYPTED]:
+        enc_data = req_data[ATTR_WEBHOOK_ENCRYPTED_DATA]
+        webhook_payload = _decrypt_payload(device[CONF_SECRET], enc_data)
+
+    if (webhook_type not in WEBHOOK_TYPES and
+            device[ATTR_APP_ID] in INTEGRATIONS_MAP):
```

Comment:
What's the use of keeping track of an `INTEGRATIONS_MAP`? We would like mobile apps to be able to distribute a custom component too.

Reply:
Okay, so what are you thinking then? App should provide the component and platform to load in the registration?

Reply:
That's what I ended up doing. `app_component` is now a valid key at registration.